aptos-move/move-examples/fungible_asset/sources/fungible_asset.move

```diff
@@ -0,0 +1,389 @@
+module fungible_asset::fungible_asset {
+    use aptos_std::smart_table::{Self, SmartTable};
+    use aptos_framework::object::{Self, Object, object_address, is_owner, generate_transfer_ref, DeleteRef, generate_delete_ref};
+    use std::signer;
+    use std::error;
+    use std::signer::address_of;
+    use std::option;
+    use std::option::Option;
+    #[test_only]
+    use aptos_framework::object::ConstructorRef;
+    #[test_only]
+    use aptos_framework::account::create_account_for_test;
+
+    friend fungible_asset::fungible_source;
+
+    /// The coin resource existence error.
+    const EFUNGIBLE_ASSET: u64 = 1;
+    /// Amount cannot be zero.
+    const EAMOUNT_CANNOT_BE_ZERO: u64 = 2;
+    /// Not the owner.
+    const ENOT_OWNER: u64 = 3;
+    /// The token account has asset mismatch.
+    const EASSET_ADDRESS_MISMATCH: u64 = 4;
+    /// The token account has positive amount so cannot be deleted.
+    const EBALANCE_NOT_ZERO: u64 = 5;
+    /// The token account is still frozen so cannot be deleted.
+    const ESTILL_FROZEN: u64 = 6;
+    /// The coin object existence error.
+    const EFUNGIBLE_ASSET_OBJECT: u64 = 7;
+    /// Insufficient amount.
+    const EINSUFFICIENT_BALANCE: u64 = 8;
+    /// FungibleAsset type mismatch.
+    const EFUNGIBLE_ASSET_TYPE_MISMATCH: u64 = 9;
+
+
+    struct FungibleAssetStore has key {
+        index: SmartTable<address, Object<FungibleAsset>>
+    }
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct FungibleAsset has key {
+        asset_addr: address,
+        balance: u64,
+    }
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct FungibleAssetProperty has key {
+        frozen: bool,
+        delete_ref: DeleteRef
+    }
+
+    // ================================================================================================================
+    // Public functions
+    // ================================================================================================================
+
+    /// Check the amount of an account.
+    public fun balance<T: key>(
```

Comment:
I think this is more like primary_balance, whereas balance would just read that of a given asset...

said another way, we should have two types of functions, those that operate on a specific object -- irrespective of source type and those that operate on a user's primary version (stored in the smartmap)

e.g.,

```
public fun balance<T: key>(fungible_asset: &Object<T>): u64 {
...
}

public fun primary_balance<T: key>(owner: address, fungible_source: &Object<T>): u64 {
...
}
```

Reply:
we could even make the notion of a primary an independent module... and opt-in

Reply:
I don't get this comment. let's talk tmr.

Reply:
unable to return reference makes this so painful.